src/main/java/org/apache/datasketches/kll/KllSketch.java

```diff
@@ -0,0 +1,1634 @@
+ * 2) After a compaction, (sum of capacities) - (sum of items) >= 1,
+ *  so there is room for least 1 more item in level zero.
+ * 3) There are no gaps except at the bottom, so if levels_[0] = 0,
+ *  the sketch is exactly filled to capacity and must be compacted or the itemsArray and levelsArray
+ *  must be expanded to include more levels.
+ * 4) Sum of weights of all retained items == N.
+ * 5) Current total item capacity = itemsArray.length = levelsArray[numLevels].
+ */
+
+/**
+ * This class is the root of the KLL sketch class hierarchy. It includes the public API that is independent
+ * of either sketch type (float or double) and independent of whether the sketch is targeted for use on the
+ * heap or Direct (off-heap.
+ *
+ * <p>Please refer to the documentation in the package-info:<br>
+ * {@link org.apache.datasketches.kll}</p>
+ *
+ * @author Lee Rhodes, Kevin Lang
+ */
+public abstract class KllSketch {
+  static final Random random = new Random();
+  static final boolean compatible = true; //rank 0.0 and 1.0. compatible with classic Quantiles Sketch
+  //final int M = DEFAULT_M; // configured minimum buffer "width", default is 8.
+  SketchType sketchType;
+  WritableMemory wmem;
+  MemoryRequestServer memReqSvr;
+  boolean direct;
+
+  /**
+   *
+   * @param sketchType either DOUBLE_SKETCH or FLOAT_SKETCH
+   * @param wmem  the current WritableMemory or null
+   * @param memReqSvr the given MemoryRequestServer to request a larger WritableMemory
+   */
+  KllSketch(final SketchType sketchType, final WritableMemory wmem, final MemoryRequestServer memReqSvr) {
+   this.sketchType = sketchType;
+   this.wmem = wmem;
+   if (wmem != null) {
+     this.direct = true;
+     this.memReqSvr = memReqSvr;
+   } else {
+     this.direct = false;
+     this.memReqSvr = null;
+   }
+  }
+
+public enum SketchType { FLOATS_SKETCH, DOUBLES_SKETCH }
+
+  //Static methods
+
+  /**
+   * Gets the approximate value of <em>k</em> to use given epsilon, the normalized rank error.
+   * @param epsilon the normalized rank error between zero and one.
+   * @param pmf if true, this function returns the value of <em>k</em> assuming the input epsilon
+   * is the desired "double-sided" epsilon for the getPMF() function. Otherwise, this function
+   * returns the value of <em>k</em> assuming the input epsilon is the desired "single-sided"
+   * epsilon for all the other queries.
+   *
+   * <p>Please refer to the documentation in the package-info:<br>
+   * {@link org.apache.datasketches.kll}</p>
+   * @return the value of <i>k</i> given a value of epsilon.
+   */
+  // constants were derived as the best fit to 99 percentile empirically measured max error in
+  // thousands of trials
+  public static int getKFromEpsilon(final double epsilon, final boolean pmf) {
+    //Ensure that eps is >= than the lowest possible eps given MAX_K and pmf=false.
+    final double eps = max(epsilon, 4.7634E-5);
+    final double kdbl = pmf
+        ? exp(log(2.446 / eps) / 0.9433)
```

Comment:
I have seen the same constants in KllHelper. better keep them in one place?

Reply:
That is a good catch.

Reply:
Fixed. Done.